mmap-alloc/src/lib.rs

```diff
@@ -411,90 +283,199 @@ impl MapAlloc {
-    unsafe fn realloc(&mut self, ptr: *mut u8, layout: Layout, new_layout: Layout) -> Result<*mut u8, AllocErr> {
+    unsafe fn realloc(
+        &mut self,
+        ptr: *mut u8,
+        layout: Layout,
+        new_layout: Layout,
+    ) -> Result<*mut u8, AllocErr> {
+        debug_assert!(
+            layout.size() != 0 && new_layout.size() != 0,
+            "realloc: size of layout and new_layout must be non-zero"
+        );
+
+        // alignment less than a page is fine because page-aligned objects are also aligned to
+        // any alignment less than a page
+        if new_layout.align() > self.pagesize {
+            return Err(AllocErr::invalid_input(
+                "cannot support alignment greater than a page",
+            ));
+        }
+
+        let old_size = next_multiple(layout.size(), self.pagesize);
+        let new_size = next_multiple(new_layout.size(), self.pagesize);
+        if old_size == new_size {
+            return Ok(ptr);
+        }
+        remap(ptr, old_size, new_layout.size(), false).ok_or(AllocErr::Exhausted {
+            request: new_layout,
+        })
+    }
+
+    #[cfg(any(target_os = "macos", windows))]
+    unsafe fn realloc(
+        &mut self,
+        ptr: *mut u8,
+        layout: Layout,
+        new_layout: Layout,
+    ) -> Result<*mut u8, AllocErr> {
+        // Adapted from the Rust standard library.
+
+        debug_assert!(
+            layout.size() > 0 && new_layout.size() > 0,
+            "realloc: size of layout and new_layout must be non-zero"
+        );
+
         // alignment less than a page is fine because page-aligned objects are also aligned to
         // any alignment less than a page
         if new_layout.align() > self.pagesize {
-            return Err(AllocErr::invalid_input("cannot support alignment greater than a page"));
+            return Err(AllocErr::invalid_input(
+                "cannot support alignment greater than a page",
+            ));
         }
-        debug_assert!(layout.size() > 0 && new_layout.size() > 0,
-                      "usable_size: size of layout and new_layout must be non-zero");
 
         let old_size = next_multiple(layout.size(), self.pagesize);
         let new_size = next_multiple(new_layout.size(), self.pagesize);
+
         if old_size == new_size {
             return Ok(ptr);
         }
-        self.realloc_helper(ptr, old_size, new_layout.size()).ok_or(AllocErr::Exhausted { request: new_layout })
+
+        if !cfg!(windows) {
+            // Windows cannot grow or shrink existing mappings.
+            if new_size >= old_size {
+                if let Ok(()) = self.grow_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            } else if new_size < old_size {
+                if let Ok(()) = self.shrink_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            }
+        }
```

Comment:
Since we're not overriding shrink_in_place for macos either, I don't think this makes sense to have. The cases which would be handled by the default implementation (and are handled for macos and windows) should already be handled by the `old_size == new_size` check above, which checks if the old layout and the new layout both round  up to the same size.

Reply:
Sorry, I hoped there would be more context in the preview. The block starts [here](https://github.com/ezrosent/allocators-rs/pull/87/files/3e5636db556f92e292af5ece2a13e21ab4809f89#diff-7be4d90297df3fdf7c0f62394837b6d5R423)

Reply:
I fixed this by implementing `shrink_in_place` on Mac and simplifying the logic here. How does that look?